lib/charms/saml_integrator/v0/saml.py

```diff
@@ -0,0 +1,303 @@
+
+    Attrs:
+        entity_id: SAML entity ID.
+        metadata_url: URL to the metadata.
+        certificates: Tuple containing the SAML certificates.
+        endpoints: Tuple containing the SAML endpoints.
+    """
+
+    @property
+    def entity_id(self) -> str:
+        """Fetch the SAML entity ID from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("entity_id")
+
+    @property
+    def metadata_url(self) -> str:
+        """Fetch the SAML metadata URL from the relation."""
+        assert self.relation.app
+        return parse_obj_as(AnyHttpUrl, self.relation.data[self.relation.app].get("metadata_url"))
+
+    @property
+    def certificates(self) -> Tuple[str, ...]:
+        """Fetch the SAML certificates from the relation."""
+        assert self.relation.app
+        return tuple(self.relation.data[self.relation.app].get("x509certs").split(","))
+
+    @property
+    def endpoints(self) -> Tuple[SamlEndpoint, ...]:
+        """Fetch the SAML endpoints from the relation."""
+        assert self.relation.app
+        relation_data = self.relation.data[self.relation.app]
+        endpoints = [
+            SamlEndpoint.from_relation_data(
+                {
+                    key2: relation_data.get(key2)
+                    for key2 in relation_data
+                    if key2.startswith("_".join(key.split("_")[:-1]))
+                }
+            )
+            for key in relation_data
+            if key.endswith("_redirect_url") or key.endswith("_post_url")
+        ]
+        endpoints.sort(key=lambda ep: ep.name)
+        return tuple(endpoints)
+
+
+class SamlRequiresEvents(ops.CharmEvents):
+    """SAML events.
+
+    This class defines the events that a SAML requirer can emit.
+
+    Attrs:
+        saml_data_available: the SamlDataAvailableEvent.
+    """
+
+    saml_data_available = ops.EventSource(SamlDataAvailableEvent)
+
+
+class SamlRequires(ops.Object):
+    """Requirer side of the SAML relation.
+
+    Attrs:
+        on: events the provider can emit.
+    """
+
+    on = SamlRequiresEvents()
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        """Construct.
+
+        Args:
+            charm: the provider charm.
+            relation_name: the relation name.
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Event emitted when the relation has changed.
+
+        Args:
+            event: event triggering this handler.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        self.on.saml_data_available.emit(event.relation, app=event.app, unit=event.unit)
```

Comment:
@arturo-seijas I have some questions regarding the `saml_data_available` event. Why is it emitted only on the leader unit when the requirer simply reads the reaction data without writing to the relation data? Additionally, considering that relation data might not always exist when the `relation-changed` event occurs, like the first `relation-changed` after `relation-joined`, shouldn't we emit the `saml_data_available` event only when the relation data is present?

Reply:
Good point. I'll be releasing a new version with these changes. Thank you